### PR TITLE
Doe add more objectives

### PR DESCRIPTION
**Pull request by Osburg**
Adds new objective functions (A-, E-, K-, G-optimality and a space filling objective). Not all of them seem to work nicely in every situation, so probably some objectives will be removed soon.

### CODE REVIEW

**Comment by Osburg** [comment]
@WaStCo I've added an objective for filling the design space homogeneously (see example [here](https://github.com/experimental-design/bofire/blob/b168825022fc8d417327430161bb0abbe037bbab/tutorials/doe/optimality_criteria.ipynb)). Is this  more or less what you wanted to have? Looking forward to feedback if it works for you as intended.

File: bofire/strategies/doe/objective.py [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+
+        return J.flatten()
+
+
+# TODO: test
+class SpaceFilling(Objective):
+
+    # TODO: how many values should be taken into account. 1 is not enough. Right now it is n_experiments, this seems to work.
+    def evaluate(self, x: np.ndarray) -> float:
+        X = self._convert_input_to_tensor(x, requires_grad=False)
```
Comment (DavidWalz) [inline bofire/strategies/doe/objective.py:423]: `_convert_input_to_tensor` computes the model matrix, space filling should be model independent, only consider (x1, x2) and not e.g. (x1, x2, x1*x2, x1**2, x2**2) for the distance calculation.

**Review by DavidWalz — COMMENTED** [review]
Awesome!

File: bofire/strategies/doe/objective.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+        return float(
+            -torch.sum(torch.sort(torch.pdist(X.detach()))[0][: self.n_experiments])
```
[the remark this answers is not in the document]
Comment (Osburg) [reply bofire/strategies/doe/objective.py:425]: Yes! If you use min (i.e. maximize the minimal distance) the optimization does not work properly. Even after 1000+ iterations almost no progress is done. My idea is that we don't "do enough" this way, since the gradient only gives information about the way one single pair of experiments should be moved (the one with the minimal distance).
Instead, we take the sum of the smallest n_experiments experiment distances as objective. This gives us information about how to move many more experiments while still only looking at some of the smallest distances (in total we have `n_experiments!` distances), so the optimum where we end up should not be disturbed to much by taking into account some more points. 
However, I only did tests until 100 experiments. So I am not sure yet if the scaling #considered_distances = O(n_experiments) is also good for much larger numbers of experiments.

File: bofire/strategies/doe/objective.py (hunk 1 above)
Comment (Osburg) [reply bofire/strategies/doe/objective.py:423]: No, `_convert_input_to_tensor()` is slightly different.
```
    def _convert_input_to_tensor(
        self, x: np.ndarray, requires_grad: bool = True
    ) -> Tensor:
        X = pd.DataFrame(
            x.reshape(len(x.flatten()) // self.n_vars, self.n_vars), columns=self.vars
        )
        return torch.tensor(X.values, requires_grad=requires_grad, **tkwargs)
```
It misses the `get_model_matrix()` call that `_convert_input_to_model_tensor()` has. But I have not forgotten that you wanted a separate class for the `SpaceFilling`! This was just a quick experiment if this objective works, will convert the PR to a draft for now.

File: bofire/strategies/doe/design.py [hunk 3]
```diff
@@ -83,11 +43,11 @@ def find_local_max_ipopt(
         tol (float): Tolerance for linear/NChooseK constraint violation. Default value is 0.
         delta (float): Regularization parameter. Default value is 1e-3.
         ipopt_options (Dict): options for IPOPT. For more information see [this link](https://coin-or.github.io/Ipopt/OPTIONS.html)
-        jacobian_building_block (Callable): Only needed for models of higher order than 3. derivatives
-            of each model term with respect to each input variable.
         sampling (Sampling, np.ndarray): Sampling class or a np.ndarray object containing the initial guess.
         fixed_experiments (pd.DataFrame): dataframe containing experiments that will be definitely part of the design.
             Values are set before the optimization.
+        objective (str): Keyword indicating which objective function to use. Valid keywords are "d" for D-optimality, "e" for E-optimality
```
[the remark this answers is not in the document]
Comment (WaStCo) [reply bofire/strategies/doe/design.py:47]: What you think of enum + default option (i.e. D-optimality).

**Comment by WaStCo** [comment]
That is great, @Osburg! This goes exactly in the direction (and the option to easily change DoE method).

Looks like uniform space filling, right? This is definetively useful!

At some point we could think about adding sequential point arranging space-filling designs (they might be interesting for Bayesian modelling in case that not all planned experiments can be performed).  @R-M-Lee , what do you think?

For example:
grafik
[source](https://www.researchgate.net/publication/289967333_Metamodels_for_optimum_design_of_composite_stiffened_structures_under_torsion_loading)

File: bofire/strategies/doe/objective.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavidWalz) [reply bofire/strategies/doe/objective.py:425]: Oh that's brilliant! 
Yes, using min will within one step only move the closest pair apart. With your formulation all points in the `n` closest pairs are moved. 
The number of distances is `n (n - 1) / 2` (not `n!`) so the scaling is fine.

File: bofire/strategies/doe/objective.py (hunk 1 above)
Comment (DavidWalz) [reply bofire/strategies/doe/objective.py:423]: True. Didn't see that `_convert_to_model_tensor` is redefined in `SpaceFilling`.

File: bofire/strategies/doe/objective.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavidWalz) [reply bofire/strategies/doe/objective.py:425]: And using the `n` closest pairs is a good choice. It's not too large, as only at the equidistant optimum, all points can simumltaneously contribute.

**Comment by jduerholt** [comment]
Hi @Osburg, this looks really great. We should then also include the new objectives as options in the `DoEStrategy`, as the user at then end should use the DoEStartegy as they provides the same interface as the other strategies in BoFire, maybe we should also update this at some point in the notebooks.

**Comment by R-M-Lee** [comment]
> That is great, @Osburg! This goes exactly in the direction (and the option to easily change DoE method).
> 
> Looks like uniform space filling, right? This is definetively useful!
> 
> At some point we could think about adding sequential point arranging space-filling designs (they might be interesting for Bayesian modelling in case that not all planned experiments can be performed). @R-M-Lee , what do you think?
> 
> For example: grafik [source](https://www.researchgate.net/publication/289967333_Metamodels_for_optimum_design_of_composite_stiffened_structures_under_torsion_loading)

I also think this is a good idea. For this, the doe strategy should accept previous observations and append these to the candidate DoE (as static values) during the optimization.

In general, this is great work. Thanks @Osburg!

**Comment by Osburg** [comment]
@R-M-Lee @WaStCo Regarding the consideration of previous observations that should be considered as fixed experiments during the optimization: we already have this feature. You can set fixed experiments using the fixed_experiments argument in `find_local_max_ipopt()` or the candidates attribute in `DoEStrategy`. Is this the kind of feature you have in mind? Or am I misunderstanding sth?

File: bofire/strategies/doe/objective.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Osburg) [reply bofire/strategies/doe/objective.py:423]: Yes, you are right. `n(n-1)/2` is of course correct. Good to hear that you agree to the objective as it is now. Then I leave things as they are rn.

**Review by DavidWalz — APPROVED** [review]
Great!

**Comment by R-M-Lee** [comment]
> @R-M-Lee @WaStCo Regarding the consideration of previous observations that should be treated as fixed experiments during the optimization: we already have this feature. You can set fixed experiments using the fixed_experiments argument in `find_local_max_ipopt()` or the candidates attribute in `DoEStrategy`. Is this the kind of feature you have in mind? Or am I misunderstanding sth?

Yes, that's what I meant at least. Ok, then this is already taken care of. Thank you!
